tezedge_state/src/proposer/mio_manager.rs

```diff
@@ -0,0 +1,279 @@
+use mio::net::{TcpListener, TcpSocket, TcpStream};
+use slab::Slab;
+use std::collections::HashMap;
+use std::io;
+use std::net::{IpAddr, Ipv4Addr, SocketAddr};
+use std::sync::Arc;
+use std::time::{Duration, Instant};
+
+use super::*;
+use crate::PeerAddress;
+
+pub type MioEvent = mio::event::Event;
+pub type NetPeer = Peer<TcpStream>;
+
+pub const MIO_SERVER_TOKEN: mio::Token = mio::Token(usize::MAX);
+pub const MIO_WAKE_TOKEN: mio::Token = mio::Token(usize::MAX - 1);
+
+impl NetworkEvent for MioEvent {
+    #[inline(always)]
+    fn is_server_event(&self) -> bool {
+        self.token() == MIO_SERVER_TOKEN
+    }
+
+    #[inline(always)]
+    fn is_readable(&self) -> bool {
+        self.is_readable()
+    }
+
+    #[inline(always)]
+    fn is_writable(&self) -> bool {
+        self.is_writable()
+    }
+
+    #[inline(always)]
+    fn is_read_closed(&self) -> bool {
+        self.is_error() || self.is_read_closed()
+    }
+
+    #[inline(always)]
+    fn is_write_closed(&self) -> bool {
+        self.is_write_closed()
+    }
+}
+
+#[derive(Debug)]
+pub struct MioEvents {
+    mio_events: mio::Events,
+    tick_event_time: Option<Instant>,
+}
+
+impl MioEvents {
+    pub fn new() -> Self {
+        Self {
+            mio_events: mio::Events::with_capacity(0),
+            tick_event_time: None,
+        }
+    }
+}
+
+impl<'a> IntoIterator for &'a MioEvents {
+    type Item = EventRef<'a, MioEvent>;
+    type IntoIter = MioEventsIter<'a>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        MioEventsIter {
+            mio_events_iter: self.mio_events.iter(),
+            tick_event_time: self.tick_event_time.clone(),
+        }
+    }
+}
+
+pub struct MioEventsIter<'a> {
+    mio_events_iter: mio::event::Iter<'a>,
+    tick_event_time: Option<Instant>,
+}
+
+impl<'a> Iterator for MioEventsIter<'a> {
+    type Item = EventRef<'a, MioEvent>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let tick_event_time = &mut self.tick_event_time;
+
+        self.mio_events_iter
+            .next()
+            .map(|event| EventRef::Network(event))
+            .or_else(|| tick_event_time.take().map(|time| Event::Tick(time)))
+    }
+}
+
+impl Events for MioEvents {
+    fn set_limit(&mut self, limit: usize) {
+        self.mio_events = mio::Events::with_capacity(limit);
+    }
+}
+
+pub struct MioManager {
+    listen_addr: SocketAddr,
+    backlog_size: u32,
+
+    poll: mio::Poll,
+    waker: Arc<mio::Waker>,
+    server: Option<TcpListener>,
+
+    address_to_token: HashMap<PeerAddress, usize>,
+    peers: Slab<NetPeer>,
+}
+
+impl MioManager {
+    const DEFAULT_BACKLOG_SIZE: u32 = 255;
+
+    pub fn new(listen_addr: SocketAddr) -> Self {
+        let poll = mio::Poll::new().expect("failed to create mio::Poll");
+        let waker = Arc::new(
+            mio::Waker::new(poll.registry(), MIO_WAKE_TOKEN).expect("failed to create mio::Waker"),
```

Comment:
@binier 
I guess, we should not use any expect/unwrap here

Reply:
We probably should. I dont think we can handle those errors anyhow.

Reply:
@binier 
I got your point, it is done just on startup, it start or it fails,
but for example if open database before that or run other stuff, we loose kind of possiblity for gracefull shutdown for example,

and all the time we are trying to remove panic/expect/unwraps (just for sure) from everywhere (ok, except the tests)

Reply:
True, but we have those expects everywhere in "light_node/main.rs" for example. We can create task for it, to remove all `unwraps` and `expects` everywhere and have graceful shutdown.